x-pack/plugins/observability_solution/apm/public/components/app/transaction_details/waterfall_with_summary/waterfall_container/waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -161,15 +161,15 @@ function getSpanItem(span: WaterfallSpan, linkedChildrenCount: number = 0): IWat
   };
 }
 
-function getErrorItem(
+export function getErrorItem(
   error: WaterfallError,
   items: IWaterfallItem[],
   entryWaterfallTransaction?: IWaterfallTransaction
 ): IWaterfallError {
   const entryTimestamp = entryWaterfallTransaction?.doc.timestamp.us ?? 0;
-  const parent = items.find((waterfallItem) => waterfallItem.id === error.parent?.id) as
-    | IWaterfallSpanOrTransaction
-    | undefined;
+  const parent = items.find(
+    (waterfallItem) => waterfallItem.id === (error.parent?.id ?? error.span?.id)
```

Comment:
@jennypavlova have you considered moving this logic to the server instead of handling it on the client? I ask that because we'll probably have many hidden problems if we support multiple schemas on the client. As @AlexanderWert has already [pointed out](https://github.com/elastic/kibana/pull/195796#issuecomment-2405460970) there's another place where the same logic is needed.

I wonder if the UI should be schema-agnostic and respect its own schema. 

This is similar to a draft PR @rmyz opened recently https://github.com/elastic/kibana/pull/194100#pullrequestreview-2333056611.

Reply:
Thanks for the comment, @cauemarcondes 
In general, I agree with you but this is just a fallback to a different field if the parent id is not there, I wouldn't call it a 'new' schema. It can be confusing if we modify the response and return the `parent.id` in a document that doesn't have it but it probably helps us in the long run as we will have the parent id available in all places in the UI. I remember that we discussed with @dgieselaar and @gregkalapos  that we wanted to do a fallback on the client to keep it compatible and return the fields in the same format as they come from elasticsearch in the server. 
But we can do it on the server, just need the changes made in https://github.com/elastic/kibana/pull/195242 as a base and merge a PR there

> @AlexanderWert has already https://github.com/elastic/kibana/pull/195796#issuecomment-2405460970 there's another place where the same logic is needed.

This is unrelated. I investigated more and this happens because the `transaction.id` is missing and the query and it is not returning the transaction object - even if I do what you suggested it won't fix the issue mentioned there

<img width="1919" alt="image" src="https://github.com/user-attachments/assets/1a4090f4-2628-40a0-b544-071d1fa88072">

It's a different query (changed [here](https://github.com/crespocarlos/kibana/blob/192606-move-source-to-fields/x-pack/plugins/observability_solution/apm/server/routes/traces/get_trace_items.ts#L103-L104)) I am not sure why we are missing the transaction id in the error document - I see that also span id is not there (using the test env)

Reply:
@jennypavlova ahhh right, great finding!

I think it's not that easy (maybe impossible) to actually have the `transaction.id` on OTel errors, because it cannot be enriched from other attributes but would be required to be collected by the SDKs / APM Agents. That's a conceptual limitation. So, then I think let's proceed with this PR as is.

\cc @gregkalapos @felixbarny correct me if I'm wrong with the above

Reply:
I used span id for Otel errors - the [PR](https://github.com/crespocarlos/kibana/pull/4/files) is merged to https://github.com/elastic/kibana/pull/195242